jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -2388,17 +2406,31 @@ public boolean equalsIgnoreCase(String string) {
 				// overloads rather than code points.
 				/*[ENDIF]*/
 				if (byteAtO1 != byteAtO2 
-						&& toUpperCase(helpers.byteToCharUnsigned(byteAtO1)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2))
-						&& toLowerCase(helpers.byteToCharUnsigned(byteAtO1)) != toLowerCase(helpers.byteToCharUnsigned(byteAtO2))) {
+						&& toUpperCase(helpers.byteToCharUnsigned(byteAtO1)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2))) {
 					return false;
 				}
 			}
 		} else {
-			while (o1 < end) {
+			// Compare the last chars.
+			// In order to tell 2 chars are different:
+			// If at least one char is ASCII, converting to upper cases then compare should be sufficient.
+			// If both chars are not in ASCII char set, need to convert to lower case and compare as well.
+			char charAtO1Last = s1.charAtInternal(s1len-1, s1Value);
+			char charAtO2Last = s2.charAtInternal(s1len-1, s2Value);
+
+			if (charAtO1Last != charAtO2Last 
+					&& toUpperCase(charAtO1Last) != toUpperCase(charAtO2Last)
+					&& (charAtO1Last < 128 || charAtO2Last < 128 || Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last))) {
+				return false;
+			}
+
+			while (o1 < end-1) {
 				char charAtO1 = s1.charAtInternal(o1++, s1Value);
 				char charAtO2 = s2.charAtInternal(o2++, s2Value);
 
-				if (charAtO1 != charAtO2 && toUpperCase(charAtO1) != toUpperCase(charAtO2) && toLowerCase(charAtO1) != toLowerCase(charAtO2)) {
+				if (charAtO1 != charAtO2 
+						&& toUpperCase(charAtO1) != toUpperCase(charAtO2) 
+						&& (charAtO1 < 128 || charAtO2 < 128 || Character.toLowerCase(charAtO1) != Character.toLowerCase(charAtO2))) {
```

Comment:
Similarly to the above, this can use 256

Reply:
Fixed.